backend/shard/src/connection.rs

```diff
@@ -0,0 +1,195 @@
+    E: std::fmt::Debug + std::fmt::Display,
+    In: serde::Serialize + Send + 'static,
+    Out: serde::de::DeserializeOwned + Send + 'static,
+{
+    use bincode::Options;
+    use soketto::handshake::{Client, ServerResponse};
+
+    let host = telemetry_uri.host().unwrap_or("127.0.0.1");
+    let port = telemetry_uri.port_u16().unwrap_or(8000);
+    let path = telemetry_uri.path();
+
+    let socket = TcpStream::connect((host, port)).await?;
+    socket.set_nodelay(true).unwrap();
+
+    // Open a websocket connection with the relemetry core:
+    let mut client = Client::new(socket.compat(), host, &path);
+    let (mut ws_to_connection, mut ws_from_connection) = match client.handshake().await? {
+        ServerResponse::Accepted { .. } => client.into_builder().finish(),
+        ServerResponse::Redirect { status_code, .. } | ServerResponse::Rejected { status_code } => {
+            return Err(anyhow::anyhow!(
+                "Failed to connect to {}{}, status code: {}",
+                host,
+                path,
+                status_code
+            ));
+        }
+    };
+
+    // This task reads data sent from the telemetry core and
+    // forwards it on to our aggregator loop:
+    tokio::spawn(async move {
+        let mut data = Vec::with_capacity(128);
+        loop {
+            // Clear the buffer and wait for the next message to arrive:
+            data.clear();
+            if let Err(e) = ws_from_connection.receive_data(&mut data).await {
+                // Couldn't receive data may mean all senders are gone, so log
+                // the error and shut this down:
+                log::error!(
+                    "Shutting down websocket connection: Failed to receive data: {}",
+                    e
+                );
+                return;
+            }
+
+            // Attempt to deserialize, and send to our handler loop:
+            match bincode::options().deserialize(&data) {
+                Ok(msg) => {
+                    if let Err(e) = tx_to_external.send(Message::Data(msg)).await {
+                        // Failure to send to our loop likely means it's hit an
+                        // issue and shut down, so bail on this loop as well:
+                        log::error!(
+                            "Shutting down websocket connection: Failed to send data out: {}",
+                            e
+                        );
+                        return;
+                    }
+                }
+                Err(err) => {
+                    // Log the error but otherwise ignore it and keep running:
+                    log::warn!("Failed to decode message from Backend Core: {:?}", err);
+                }
+            }
+        }
+    });
+
+    // This task receives messages from the aggregator,
+    // encodes them and sends them to the telemetry core:
+    let (tx_to_connection, mut rx_from_aggregator) = mpsc::channel(10);
```

Comment:
what's the rationale behind setting channel size to 10?

Reply:
Nothing very tangible; I think I read that the default actix channel size was 10 (for the previous telemetry impl) so I thought I'd start there (though I also think we primarily ignored that size by using `do_send` to send messages).

Really I need to benchmark this and figure out what the channel size should be or whether some channels should be unbounded (but I was keen to allow for some backpressure for now)

Reply:
ok, cool we were bitten quite badly about this in [jsonrpsee previously with a deadlock w.r.t bounded channels](https://github.com/paritytech/jsonrpsee/issues/143) that's the reason I'm asking :)

Reply:
Thanks for the pointer! I have pondered disconnecting nodes if the shard can't keep up (that may also allow the next node connection to be load balanced onto a shard that's under less stress). Currently, outgoing connections to feeds are unbounded (but I can def see a case for doing something smarter here and dropping messages if the buffer grows too much; I don't like the idea that a feed can be deliberately slow and cause issues that way!

Definitely more thought is needed in this area!